### PR TITLE
Magic Links Schema

### PR DESCRIPTION
**Pull request by jimleeder123**
The main difference with this model is that the timestamps are disabled. Otherwise this is the setup of a migration, model, factory and a relationship on the user model.

### CODE REVIEW

**Comment by jimleeder123** [comment]
[Magic links schema](https://trello.com/c/1OedjdeD/68-magic-links-schema)

File: app/Models/MagicLink.php [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+use Illuminate\Database\Eloquent\Model;
+use Illuminate\Database\Eloquent\Relations\BelongsTo;
+
+/**
+ * @property int $id
+ * @property string $uuid
+ * @property int $user_id
+ * @property CarbonImmutable $requested_at
+ * @property CarbonImmutable $expires_at
+ * @property CarbonImmutable $authenticated_at
```
[the remark this answers is not in the document]
Comment (jimleeder123) [reply app/Models/MagicLink.php:17]: Done

File: app/Models/MagicLink.php [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+ * @property string $ip_requested_from
+ * @property string|null $ip_authenticated_from
+ * @property string $intended_url
```
[the remark this answers is not in the document]
Comment (jimleeder123) [reply app/Models/MagicLink.php:20]: Done

File: app/Models/MagicLink.php [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+     */
+    protected $table = 'magic_links';
+
+    /**
+     * The attributes that should be cast.
+     *
+     * @var array<string, string|class-string>
+     */
+    protected $casts = [
+        // todo: discuss - do we need to cast IP addresses?
```
[the remark this answers is not in the document]
Comment (jimleeder123) [reply app/Models/MagicLink.php:42]: Thanks, I will leave the casts as an empty array

File: app/Models/MagicLink.php [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        return $this->belongsTo(User::class, 'user_id', 'id');
+    }
+
+    /**
+     * @return bool
+     */
+    public function hasAuthenticated(): bool
+    {
+        return $this->authenticated_at->isPast();
+    }
```
[the remark this answers is not in the document]
Comment (jimleeder123) [reply app/Models/MagicLink.php:61]: Removed

File: app/Models/MagicLink.php [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+    public function hasExpired(): bool
+    {
+        return $this->expires_at->isPast() && $this->authenticated_at == null;
+    }
```
[the remark this answers is not in the document]
Comment (jimleeder123) [reply app/Models/MagicLink.php:66]: I think it is needed - but there is perhaps another situation a magic link can expire. This is what I think could happen...

1. Expires and the user authenticates
2. Expires but the user never authenticates (this is what the function above is for)

Is it worth adding a function for the first scenario too? I don't know how much we'll need them but it's useful and preparing for the rest of the build?

File: app/Models/User.php [hunk 6]
```diff
@@ -66,4 +66,12 @@ public function enquiries(): HasMany
     {
         return $this->hasMany(Enquiry::class, 'user_id');
     }
+
+    /**
+     * @return HasMany<MagicLink>
+     */
+    public function magicLinks(): HasMany
```
[the remark this answers is not in the document]
Comment (jimleeder123) [reply app/Models/User.php:73]: Done

File: database/migrations/2024_02_08_153045_create_magic_links_table.php [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    {
+        Schema::create('magic_links', function (Blueprint $table) {
+            $table->id();
+            $table->uuid()->unique();
+            $table->foreignId('user_id')->references('id')->on('users');
+            $table->dateTime('requested_at');
+            $table->datetime('expires_at');
+            $table->datetime('authenticated_at')->nullable();
+            $table->ipAddress('ip_requested_from');
+            $table->ipAddress('ip_authentication_from')->nullable();
```
[the remark this answers is not in the document]
Comment (jimleeder123) [reply database/migrations/2024_02_08_153045_create_magic_links_table.php:22]: Done

File: database/factories/MagicLinkFactory.php [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    {
+        $ipAddress = $this->faker->ipv4;
+
+        return [
+            'user_id' => User::factory(),
+            'requested_at' => CarbonImmutable::now()->subMinutes(5),
+            'expires_at' => CarbonImmutable::now()->addMinutes(5),
+            'authenticated_at' => null,
+            'ip_requested_from' => $ipAddress,
+            'ip_authentication_from' => $ipAddress,
```
[the remark this answers is not in the document]
Comment (jimleeder123) [reply database/factories/MagicLinkFactory.php:37]: Done

File: database/factories/MagicLinkFactory.php [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+            'intended_url' => route('web.space-calculator.index'),
+            // todo: discuss - should we use a different route for intended url default?
```
[the remark this answers is not in the document]
Comment (jimleeder123) [reply database/factories/MagicLinkFactory.php:39]: Left it as null

**Comment by jimleeder123** [comment]
> Looks good, just a few small detail bits to tweak.

Thanks, this is ready for review again

**Review by gbuckingham89 — APPROVED** [review]
Looks good to go! ✅

**Pull request merged**